DataTransformation/SqlServer/SqlDataSource.cs

```diff
@@ -3,45 +3,107 @@
+            }
+        }
+
+        public string ProcedureName
+        {
+            get
+            {
+                return _commandText;
+            }
+            set
+            {
+                _commandType = CommandType.StoredProcedure;
+                _commandText = value;
+            }
         }
 
-        public string CommandText { get; set; }
         public ParameterSet Parameters { get; }
         public uint Timeout { get; set; }
 
+        public void MapColumns<T>(string sourceName, string outputAlias)
+        {
+            _columnMappings.Add(new DataElementPairing(sourceName, outputAlias, typeof(T)));
+        }
+
         public override DataTable GetResults()
         {
-            return this.Execute(CommandText, CommandType.Text, SqlActivityExtensionMethods.GetDataTable);
+            DataTable rawDataTable = this.Execute(_commandText, _commandType, SqlActivityExtensionMethods.GetDataTable);
+            if (_columnMappings.Any())
+            {
+                return rawDataTable.RepackageTable(_columnMappings);
+            }
+            return rawDataTable;
         }
 
         public override T GetScalar<T>()
         {
-            return (T) this.Execute(CommandText, CommandType.Text, x => x.ExecuteScalar());
+            if (_columnMappings.Any())
+            {
+                throw new InvalidOperationException();
+            }
+            return (T) this.Execute(_commandText, _commandType, x => x.ExecuteScalar());
         }
 
         protected override IDataRecordIterator CreateReader()
         {
-            SqlDataReader reader = this.Execute(CommandText, CommandType.Text, x => x.ExecuteReader());
-            return new SqlRecordIterator(reader);
+            return new SqlRecordIterator(this, _commandText, _commandType, _columnMappings);
         }
 
         private class SqlRecordIterator : IDataRecordIterator
         {
-            private readonly SqlDataReader _reader;
+            private readonly ResourceStack _sqlPipeline;
+            private readonly List<DataElementPairing> _columnNameMappings;
+            private readonly Dictionary<int, DataElementPairing> _fieldMappings;
 
-            public SqlRecordIterator(SqlDataReader reader)
+            public SqlRecordIterator(ISqlActivity sqlSettings, string commandText, CommandType commandType, List<DataElementPairing> columnNameMappings)
             {
-                _reader = reader;
+                _sqlPipeline = new ResourceStack();
+                _columnNameMappings = columnNameMappings;
+                _fieldMappings = new Dictionary<int, DataElementPairing>();
+
+                SqlConnection connection = sqlSettings.GetConnection();
+                _sqlPipeline.Push(connection);
+
+                SqlCommand command = sqlSettings.BuildCommand(connection, commandText, commandType);
+                _sqlPipeline.Push(command);
+
+                SqlDataReader reader = command.ExecuteReader();
```

Comment:
I see the dispose below, but do not see where your closing the reader. It may be done in the Resource Stack, if so ignore this.

Reply:
Yes, that is the purpose of `ResourceStack` - to dispose of a stack of `IDisposable` objects.